core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1130,20 +1134,28 @@ protected MutableRel invert(MutableRel model, MutableRel input,
       }
       final List<RexNode> exprList = new ArrayList<>();
       final RexBuilder rexBuilder = model.cluster.getRexBuilder();
-      for (RelDataTypeField field : model.rowType.getFieldList()) {
-        exprList.add(rexBuilder.makeZeroLiteral(field.getType()));
+
+      // Initialize "exprList" with nulls.
+      for (int i = 0; i < model.rowType.getFieldCount(); i++) {
+        exprList.add(null);
       }
+      // Fill in "exprList" with list of "RexInputRef" by
+      // analyzing mapping information from "project".
       for (Ord<RexNode> expr : Ord.zip(project.projects)) {
         if (expr.e instanceof RexInputRef) {
           final int target = ((RexInputRef) expr.e).getIndex();
-          exprList.set(target,
-              rexBuilder.ensureType(expr.e.getType(),
-                  RexInputRef.of(expr.i, input.rowType),
-                  false));
-        } else {
-          throw MatchFailed.INSTANCE;
+          if (exprList.get(target) == null) {
+            exprList.set(target,
+                rexBuilder.ensureType(expr.e.getType(),
+                    RexInputRef.of(expr.i, input.rowType),
+                    false));
+          }
         }
       }
+      // If a column failed to be projected from "input", matching fails.
```

Comment:
What if there is a literal in the project? The literal doesn't reference any column in the input.

Reply:
"What if there is a literal in the project ?"
I think it doesn't matter.
When arriving this method, the new `Filter` has the `Project` of target as its input. What we want to do at this method is to add a `Project` on top of the new `Filter` and ensure the new `Filter` has the same schema as the old `Filter`. So we only take `RexInputRef` into consideration and we just neglect when it's a literal.